src/App/Fossa/VSI/Types.hs

```diff
@@ -0,0 +1,79 @@
+module App.Fossa.VSI.Types (
+  Locator (..),
+  LocatorParseError (..),
+  parseLocator,
+  isUserDefined,
+  userDefinedFetcher,
+  toDependency,
+) where
+
+import Control.Effect.Diagnostics (ToDiagnostic, renderDiagnostic)
+import Data.Text (Text)
+import DepTypes (DepType (..), Dependency (..), VerConstraint (CEq))
+import Effect.Logger (Pretty (pretty), viaShow)
+import Srclib.Converter (fetcherToDepType)
+import Srclib.Types qualified as Srclib
+
+-- | VSI supports a subset of possible Locators.
+-- | Specifically, all VSI locators must have a valid revision.
+data Locator = Locator
+  { locatorFetcher :: Text
+  , locatorProject :: Text
+  , locatorRevision :: Text
+  }
+  deriving (Eq, Ord, Show)
+
+parseLocator :: Text -> Either LocatorParseError Locator
+parseLocator a = do
+  let plain = Srclib.parseLocator a
+  validateLocator plain
+
+newtype LocatorParseError = RevisionRequired Srclib.Locator
+  deriving (Eq, Ord, Show)
+
+instance ToDiagnostic LocatorParseError where
+  renderDiagnostic (RevisionRequired locator) =
+    "Revision is required on locator: " <> viaShow locator
+
+toDependency :: Locator -> Either ToDependencyError Dependency
+toDependency locator =
+  Dependency
+    <$> validateDepType locator
+    <*> Right (locatorProject locator)
+    <*> (Right . Just . CEq $ locatorRevision locator)
+    <*> Right []
+    <*> Right []
+    <*> Right mempty
```

Comment:
[no action] I like the use of the explicit `Right` constructor here over `pure` -- I wonder if we should add that to the style guide for simple applicatives like `Either e` and `Maybe`. Not sure how often this comes up, though

cc: @scruffystuffs

Reply:
It's worth noting, though I wouldn't make it strict.  I'll add this.